src/model/selfgrav_hydrostatic_col.h

```diff
@@ -0,0 +1,538 @@
+#ifndef SELFGRAV_HYDROSTATIC_COL
+#define SELFGRAV_HYDROSTATIC_COL
+
+#include <array>
+#include <cassert>
+#include <cmath>
+#include <cstdio>
+#include <limits>
+#include <type_traits>
+
+#include "../global/global.h"
+#include "../utils/error_handling.h"
+
+namespace ode_detail
+{
+
+struct NullFn {
+  /// acts as a dummy placeholder for representing a "logger function"
+  template <int N>
+  void operator()(Real x, Real cur_step, std::array<Real, N> yvec, std::array<Real, N> cur_yvec_step) const noexcept
+  {
+  }
+};
+
+}  // namespace ode_detail
+
+template <int N, class DerivFn, class LogFn = ode_detail::NullFn>
+class ODEIntegrator
+{
+  /// Represents an integrator of a system of (coupled) 1D ODEs.
+  ///
+  /// @tparam N the number of ODEs that are integrated
+  /// @tparam DerivFn Specifies the derivative of the ODEs
+  /// @tparam LogFn An optional logger function that can be used to log
+  ///     information as the integrator progresses.
+
+ public:
+  ODEIntegrator() = delete;
+
+  ODEIntegrator(DerivFn deriv_fn, LogFn log_fn = ode_detail::NullFn{}) : deriv_fn_(deriv_fn), log_fn_(log_fn) {}
+
+  std::array<Real, N> integrate(Real xstart, Real xend, Real nominal_step,
+                                std::array<Real, N> yvec_start) const noexcept
+  {
+    Real cur_x                   = xstart;
+    std::array<Real, N> cur_yvec = yvec_start;
+
+    while (true) {
+      Real cur_step = std::min(xend - cur_x, nominal_step);
+
+      // calculate the change in yvec over cur_step
+      std::array<Real, N> d_yvec = calc_midpoint_step(cur_x, cur_yvec, cur_step);
+      // std::array<Real, N> d_yvec = calc_rk4_step(cur_x, cur_yvec, cur_step);
+
+      // potentially call the logger function
+      if constexpr (not std::is_same_v<LogFn, ode_detail::NullFn>) {
+        log_fn_(cur_x, cur_step, cur_yvec, d_yvec);
+      }
+
+      // update cur_x and cur_yvec
+      cur_x += cur_step;
+      for (int i = 0; i < N; i++) {
+        cur_yvec[i] += d_yvec[i];
+      }
+
+      if (cur_step < nominal_step) {
```

Comment:
logic might need to be `cur_step <= nominal_step` to avoid potential extra loop.

Reply:
Thanks for catching this. I fixed this (the needed logic was actually slightly different)